csharp/src/Ice/SocketStream.cs

```diff
@@ -228,17 +300,24 @@ internal async ValueTask<(IncomingRequestFrame, bool)> ReceiveRequestFrameAsync(
                 _socket.TraceFrame(Id, request);
             }
 
-            return (request, fin);
+            if (!ReceivedEndOfStream)
```

Comment:
I find this code not very elegant. You could make this check prior to the construction of IncomingRequestFrame, and pass the stream to the ctor. This way, the SocketStream property (which does not need to be on the base class) could be get-only.

Also, do we really need this _useCount? It seems that once the create an IncomingRequest/ResponseFrame with a stream, the frame is responsible to dispose the stream. Who else is going to use this stream?

Reply:
I didn't want to make SocketStream a parameter of the constructor to not expose it to our public API (right now at least, the user has no way to access the SocketStream object). I can add an internal constructor however instead with this additional parameter.

The frame is not responsible to dispose the stream, it's either the IO steam or IAsyncEnumerable which is created after reading the incoming response/request parameters that is responsible for disposing of the stream (either when it's disposed or when EOS is reached).